### PR TITLE
feat(@desktop/wallet): cpp: Added new apis for wallet

### PR DESCRIPTION
**Pull request by Khushboo-dev-cpp**
feat(@desktop/wallet): cpp: Added new apis for wallet

### What does the PR do

adds new api in c++ for
1. Generate new API
2. Import accounts by PK
3. Import account by seed phrase
4. Add watch only address
5. delete account
6. current account

### Affected areas

Wallet

### Screenshot of functionality

https://user-images.githubusercontent.com/60327365/155000666-8f680f3d-bb5b-4bec-b8d9-29a336febe2b.mov

### CODE REVIEW

File: src-cpp/app/modules/main/wallet/accounts/controller.cpp [hunk 1]
```diff
@@ -22,4 +22,45 @@ QList<Wallets::WalletAccountDto> Controller::getWalletAccounts()
 
     return wallet_accounts;
 }
+
+QString Controller::generateNewAccount(QString password, QString accountName, QString color)
+{
+    if(m_walletServicePtr != nullptr)
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/controller.cpp:28]: `minor` no need to check it like `== or !=` just `if (m_walletServicePtr) or (!m_walletServicePtr)`

File: src-cpp/app/modules/main/wallet/accounts/controller.cpp [hunk 2]
```diff
@@ -22,4 +22,45 @@ QList<Wallets::WalletAccountDto> Controller::getWalletAccounts()
 
     return wallet_accounts;
 }
+
+QString Controller::generateNewAccount(QString password, QString accountName, QString color)
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/controller.cpp:26]: and again `const QString&`should be everywhere

File: src-cpp/app/modules/main/wallet/accounts/controller.cpp [hunk 3]
```diff
@@ -22,4 +22,45 @@ QList<Wallets::WalletAccountDto> Controller::getWalletAccounts()
...
+    }
+}
+
+QString Controller::addWatchOnlyAccount(QString address, QString accountName , QString color)
+{
+    if(m_walletServicePtr != nullptr)
+    {
+        return m_walletServicePtr->addWatchOnlyAccount(address, accountName, color);
+    }
+}
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/controller.cpp:56]: what happens if `m_walletServicePtr` is `nullptr`? which value returned?

File: src-cpp/app/modules/main/wallet/accounts/controller.cpp [hunk 4]
```diff
@@ -22,4 +22,45 @@ QList<Wallets::WalletAccountDto> Controller::getWalletAccounts()
...
+
+QString Controller::addAccountsFromSeed(QString seedPhrase, QString password, QString accountName, QString color)
+{
+    if(m_walletServicePtr != nullptr)
+    {
+        return m_walletServicePtr->addAccountsFromSeed(seedPhrase, password, accountName, color);
+    }
+}
+
+QString Controller::addWatchOnlyAccount(QString address, QString accountName , QString color)
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/controller.cpp:50]: minor: extra space `QString accountName , QString color`

File: src-cpp/app/modules/main/wallet/accounts/item.cpp [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 namespace Modules::Main::Wallet::Accounts
 {
-Item::Item(QString name, QString address, QString path, QString color, QString publicKey, QString walletType, bool isWallet, bool isChat, float currencyBalance)
+Item::Item(QString name, QString address, QString path, QString color, QString publicKey, QString walletType, bool isWallet, bool isChat, float currencyBalance, QObject* parent)
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/item.cpp:5]: `const QString&`?

File: src-cpp/app/modules/main/wallet/accounts/item.cpp [hunk 6]
```diff
@@ -59,4 +60,27 @@ float  Item::getCurrencyBalance() const
     return m_currencyBalance;
 }
 
+void Item::setData(Item* item)
+{
+    m_name = item->getName();
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/item.cpp:65]: what if `item` is null?

File: src-cpp/app/modules/main/wallet/accounts/item.h [hunk 7]
```diff
@@ -1,12 +1,24 @@
 #ifndef WALLET_ACCOUNT_ITEM_H
 #define WALLET_ACCOUNT_ITEM_H
 
+#include <QObject>
 #include <QString>
 
 namespace Modules::Main::Wallet::Accounts
 {
-class Item
+class Item: public QObject
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/item.h:9]: please, use next order:
```
class Name {
  // qt properties
  // public 
  // public slots
  // signals
  // protected
  // private methods
  // private fields
}
```

File: src-cpp/app/modules/main/wallet/accounts/model.cpp [hunk 8]
```diff
@@ -34,35 +34,40 @@ QVariant Model::data(const QModelIndex& index, int role) const
...
     {
-    case Name: return QVariant(item.getName());
-    case Address: return QVariant(item.getAddress());
-    case Path: return QVariant(item.getPath());
-    case Color: return QVariant(item.getColor());
-    case PublicKey: return QVariant(item.getPublicKey());
-    case WalletType: return QVariant(item.getWalletType());
-    case IsWallet: return QVariant(item.getIsWallet());
-    case IsChat: return QVariant(item.getIsChat());
+    case Name: return QVariant(item->getName());
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/model.cpp:46]: `minor`: no need to make it `QVariant` it's works without QVariant constructor

File: src-cpp/app/modules/main/wallet/accounts/model.cpp [hunk 9]
```diff
@@ -34,35 +34,40 @@ QVariant Model::data(const QModelIndex& index, int role) const
...
         //    case Assets: return QVariant(item.ge());
-    case CurrencyBalance: return QVariant(item.getCurrencyBalance());
+    case CurrencyBalance: return QVariant(item->getCurrencyBalance());
     }
 
     return QVariant();
 }
 
-void Model::setItems(QVector<Item> &items)
+void Model::setItems(QVector<Item*> items)
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/model.cpp:61]: `const QVector<Item*>&`

File: src-cpp/app/modules/main/wallet/accounts/model.cpp [hunk 10]
```diff
@@ -34,35 +34,40 @@ QVariant Model::data(const QModelIndex& index, int role) const
...
 {
     beginResetModel();
     m_items = items;
     endResetModel();
 }
 
+Item* Model::getItemByIndex(int index) const
+{
+    return m_items.at(index);
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/model.cpp:70]: what if index out of range?

File: src-cpp/app/modules/main/wallet/accounts/view.cpp [hunk 11]
```diff
@@ -4,14 +4,17 @@
 
 namespace Modules::Main::Wallet::Accounts
 {
-View::View(QObject* parent)
-    : QObject(parent)
+View::View(Controller *controller, QObject* parent)
+    : m_controllerPtr(controller),
+      QObject(parent)
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/view.cpp:9]: `minor`: that line should be first cos its base class constructor.

File: src-cpp/app/modules/main/wallet/accounts/view.cpp [hunk 12]
```diff
@@ -4,14 +4,17 @@
...
 {
     m_modelPtr = new Model(this);
+    m_currentAccountPtr = new Item("", "", "", "", "", "", false, false, 0);
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/view.cpp:12]: just make constructor with default parameters.

File: src-cpp/app/modules/main/wallet/accounts/view.cpp [hunk 13]
```diff
@@ -20,8 +23,120 @@ Model* View::getModel()
...
+void View::refreshWalletAccounts()
+{
+    auto walletAccounts = m_controllerPtr->getWalletAccounts();
+
+    if(walletAccounts.size() > 0)
+    {
+        QVector<Item*> items;
+        foreach(const auto& acc, walletAccounts)
+        {
+            items << new Item(acc.name, acc.address, acc.path, acc.color, acc.publicKey, acc.walletType, acc.isWallet, acc.isChat, 0, this);
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/view.cpp:41]: you can create a factory method like `Item::createFromAccount(walletAccount)`

File: src-cpp/app/modules/main/wallet/accounts/view.cpp [hunk 14]
```diff
@@ -20,8 +23,120 @@ Model* View::getModel()
...
+    }
+    else {
+        qWarning()<<"controller pointer is null";
+    }
+}
+
+void View::switchAccount(int index)
+{
+    m_currentAccountPtr->setData(m_modelPtr->getItemByIndex(index));
+    emit currentAccountChanged();
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/view.cpp:134]: what if index out of range?

File: src-cpp/app/modules/main/wallet/accounts/view.cpp [hunk 15]
```diff
@@ -20,8 +23,120 @@ Model* View::getModel()
...
+}
+
+Item* View::getCurrentAccount()
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/view.cpp:137]: `const` method.

File: src-cpp/app/modules/main/wallet/accounts/view.h [hunk 16]
```diff
@@ -5,30 +5,46 @@
...
 private:
+    void refreshWalletAccounts();
+
     Model* m_modelPtr;
+    Controller* m_controllerPtr;
+    Item* m_currentAccountPtr;
 
 public slots:
     Model* getModel();
+    Item* getCurrentAccount();
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/view.h:42]: why slots return something? if you call it by `connection` its useless, but if you use it in QML - it should be just `Q_INVOKABLE`

File: src-cpp/app_service/service/wallet_accounts/service.cpp [hunk 17]
```diff
@@ -36,9 +37,65 @@ void Service::fetchAccounts()
...
+
 QList<WalletAccountDto> Service::getWalletAccounts()
 {
     return m_walletAccounts.values();
 }
 
+QString Service::generateNewAccount(QString password, QString accountName, QString color)
+{
+    auto response = Backend::Wallet::Accounts::generateNewAccount(password, accountName, color);
+    if(response.m_error.m_message == "")
```
Comment (borismelnik) [inline src-cpp/app_service/service/wallet_accounts/service.cpp:55]: `isEmpty()`?

File: src-cpp/backend/include/backend/types.h [hunk 18]
```diff
@@ -24,10 +24,12 @@ struct RpcException : public std::exception
 class RpcError
 {
 public:
-	int m_code;
+    double m_code;
 	QString m_message;
 
 	friend ostream& operator<<(ostream& os, Backend::RpcError& r);
+    RpcError() = default;
+    RpcError(int code, QString message);
```
Comment (borismelnik) [inline src-cpp/backend/include/backend/types.h:32]: so `double` or `int`?

File: src-cpp/backend/include/backend/wallet_accounts.h [hunk 19]
```diff
@@ -5,9 +5,20 @@
...
 
-namespace Backend::Wallet::Accounts
+namespace Backend
 {
-Backend::RpcResponse<QJsonArray> getAccounts();
-} // Backend::Wallet::Accounts
+namespace Wallet
+{
+namespace  Accounts
+{
```
Comment (borismelnik) [inline src-cpp/backend/include/backend/wallet_accounts.h:13]: `namespaces`?

File: src-cpp/backend/wallet_accounts.cpp [hunk 20]
```diff
@@ -13,9 +13,63 @@ RpcResponse<QJsonArray> getAccounts()
...
 
+RpcResponse<QString> generateNewAccount(QString password, QString accountName, QString color)
+{
+    QString hashedPassword(Backend::Utils::hashString(password));
+    QJsonArray payload = {hashedPassword, accountName, color};
+    QJsonObject inputJSON{{"jsonrpc", "2.0"}, {"method", "accounts_generateAccount"}, {"params", payload}};
+    auto result = CallPrivateRPC(Utils::jsonToStr(inputJSON).toUtf8().data());
+    auto response = QJsonDocument::fromJson(result);
+    return  RpcResponse<QString>(result,response["result"].toString(),
+            RpcError(response["error"]["code"].toDouble(), response["error"]["message"].toString()));
```
Comment (borismelnik) [inline src-cpp/backend/wallet_accounts.cpp:29]: what if those properties not exists?

Comment (borismelnik) [reply src-cpp/backend/wallet_accounts.cpp:29]: so double conversion  - from `double` to `int` and from `int` again to `double`. Are you sure error code type is `double`?

File: ui/imports/utils/Constants.qml [hunk 21]
```diff
@@ -221,4 +221,6 @@ QtObject {
         //% "Continuing will require a transaction to connect the username with your current chat key."
         "connected-different-key": qsTrId("ens-username-connected-with-different-key"),
     }
+
+    readonly property bool isCppApp: Qt.application.name === "Status Desktop C++"
```
Comment (borismelnik) [inline ui/imports/utils/Constants.qml:225]: please, you can provide a singleton from C++ to QML, but checking the name of app is too unstable.

**Review by borismelnik — CHANGES_REQUESTED** [review]
I found a several topics which we can improve here. The best way to keep in mind it:
- Every `complex` (not basic types) input parameter should be  passed as `const Type&`
- Every `complex` return type can be `const Type&` (except temporary objects which can be moved)
- Every `getter` should be a const.
- `nullptr` convertible to `bool` safety 
- Every pointer as input parameters should be check for `nullptr`
- Every non-void function must return something anyway (can be fixed by compiler parameter )
- Avoid unnecessary type conversions (for example, from double to int).

This is a `basic` and `necessary` requirements to efficient and stable C++ code. We can skip some `super-new-standard-features`, but that list is `must be`.

**Comment by borismelnik** [comment]
@Khushboo-dev-cpp anyway its a great job!

File: src-cpp/app/modules/main/wallet/accounts/controller.cpp (hunk 3 above)
Comment (Khushboo-dev-cpp) [reply src-cpp/app/modules/main/wallet/accounts/controller.cpp:56]: will return an error string, thanks!

File: src-cpp/app/modules/main/wallet/accounts/controller.cpp (hunk 2 above)
Comment (Khushboo-dev-cpp) [reply src-cpp/app/modules/main/wallet/accounts/controller.cpp:26]: can't be as I am creating a QString in local scope now to handle the nullptr use case.

File: src-cpp/app/modules/main/wallet/accounts/model.cpp (hunk 10 above)
Comment (Khushboo-dev-cpp) [reply src-cpp/app/modules/main/wallet/accounts/model.cpp:70]: put it in now.

File: src-cpp/app/modules/main/wallet/accounts/view.cpp (hunk 12 above)
Comment (Khushboo-dev-cpp) [reply src-cpp/app/modules/main/wallet/accounts/view.cpp:12]: Wouldn't that affect the places that we need to call properly? like someone may miss giving some value to parameters and no error would be displayed.?

File: src-cpp/app/modules/main/wallet/accounts/view.cpp (hunk 13 above)
Comment (Khushboo-dev-cpp) [reply src-cpp/app/modules/main/wallet/accounts/view.cpp:41]: Would have to expose the DTO type to item in this case? is it really needed?

File: src-cpp/app/modules/main/wallet/accounts/view.h (hunk 16 above)
Comment (Khushboo-dev-cpp) [reply src-cpp/app/modules/main/wallet/accounts/view.h:42]: > I found a several topics which we can improve here. The best way to keep in mind it:
> 
> * Every `complex` (not basic types) input parameter should be  passed as `const Type&`
> * Every `complex` return type can be `const Type&` (except temporary objects which can be moved)
> * Every `getter` should be a const.
> * `nullptr` convertible to `bool` safety
> * Every pointer as input parameters should be check for `nullptr`
> * Every non-void function must return something anyway (can be fixed by compiler parameter )
> * Avoid unnecessary type conversions (for example, from double to int).
> * Add default parameters to Constructors.
> * Try to use build-in methods like `isEmpty` or `isValid`.
> * Check `out-of-range` for containers.
> 
> This is a `basic` and `necessary` requirements to efficient and stable C++ code. We can skip some `super-new-standard-features`, but that list is `must be`.

should create a cpp code checklist

File: src-cpp/backend/wallet_accounts.cpp (hunk 20 above)
Comment (Khushboo-dev-cpp) [reply src-cpp/backend/wallet_accounts.cpp:29]: I printed it and found it to be like -323121212 something like this

Comment (Khushboo-dev-cpp) [reply src-cpp/backend/wallet_accounts.cpp:29]: > what if those properties not exists?

so the thing is in this specific case it does exist, this sadly we need to depend on the status_go response and it is filled differently for different apis.`
And for each of them we need to add it this way.

**Comment by Khushboo-dev-cpp** [comment]
all done @borismelnik ! addressed all review comments

File: ui/imports/utils/Constants.qml (hunk 21 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/utils/Constants.qml]: as discussed exposed via context property for now. :)

File: src-cpp/app/modules/main/wallet/accounts/controller.cpp (hunk 2 above)
Comment (borismelnik) [reply src-cpp/app/modules/main/wallet/accounts/controller.cpp:28]: but what about `QString password, QString accountName, QString color`?

File: src-cpp/app/modules/main/wallet/accounts/model.h [hunk 22]
```diff
@@ -34,10 +34,11 @@ class Model : public QAbstractListModel
     QHash<int, QByteArray> roleNames() const override;
     int rowCount(const QModelIndex&) const override;
     QVariant data(const QModelIndex& index, int role) const override;
-    void setItems(QVector<Item>& items);
+    void setItems(QVector<Item *>& items);
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/model.h:37]: if you define `QVector<Item *>& items` that means you will change that parameter in method. If you use `const` reference you avoid unnecessary copy.

File: src-cpp/app/modules/main/wallet/accounts/module.h [hunk 23]
```diff
@@ -36,6 +26,15 @@ public slots:
 
 signals:
     void loaded() override;
+
+private:
+    void connect();
```
Comment (borismelnik) [inline src-cpp/app/modules/main/wallet/accounts/module.h:31]: oh please rename it cos `QObject::connect`. as a variant `void doConnect`. but not necessary :)

**Review by borismelnik — CHANGES_REQUESTED** [review]
Great!
Just last changes - input parameters of function should be `const Type&` if don't want to change it.

File: src-cpp/app/modules/main/wallet/accounts/model.h (hunk 22 above)
Comment (Khushboo-dev-cpp) [reply src-cpp/app/modules/main/wallet/accounts/model.h]: actually const won't avoid copying but its a better practice. have made it const now :)

**Review by saledjenic — APPROVED** [review]
@Khushboo-dev-cpp all what @borismelnik said is correct, and of course there is much more we should go through and align on.

File: src-cpp/app/modules/main/wallet/accounts/controller.cpp (hunk 2 above)
Comment (Khushboo-dev-cpp) [reply src-cpp/app/modules/main/wallet/accounts/controller.cpp:28]: done!

File: src-cpp/app/modules/shared/section_item.h [hunk 24]
```diff
@@ -39,15 +40,16 @@ class SectionItem : public QObject
...
-                QString icon,
-                QString color,
+    SectionItem(QObject* parent = nullptr,
+                QString id = "",
+                SectionType sectionType = SectionType::Unkown,
+                QString name = "",
+                QString description = "",
+                QString image = "",
+                QString icon = "",
+                QString color = "",
```
Comment (borismelnik) [inline src-cpp/app/modules/shared/section_item.h:50]: @Khushboo-dev-cpp sorry I think you missed that place :)

**Review by borismelnik — APPROVED** [review]
@Khushboo-dev-cpp great! just fix last place please :)

File: src-cpp/app/modules/shared/section_item.h (hunk 24 above)
Comment (Khushboo-dev-cpp) [reply src-cpp/app/modules/shared/section_item.h]: all done and merged , thanks @borismelnik !

**Pull request merged**